_drafts/making-backend-scalable.markdown

```diff
@@ -0,0 +1,130 @@
+
+```
+def TangoDictionary(object_name):
+    if Config.USE_REDIS:
+        return TangoRemoteDictionary(object_name)
+    elif Config.USE_MONGO:
+        return TangoMongoDictionary(object_name)
+    else:
+        return TangoNativeDictionary()
+```
+
+
+
+### Serialization and Marshalling ###
+
+Redis, being a key-value store, only accepts strings as values. Even though the objects we are trying to store are quite simple and flat, we do not want to create a new key-value pair for every attribute of the class. Instead we decided to serialize what we call the 'remote objects' using the `pickle` module included in the standard python library.
+
+A sample method to put the object into the store looks like
+
+```
+def set(self, id, obj):
+    pickled_obj = pickle.dumps(obj)
+    self.r.hset(self.hash_name, str(id), pickled_obj)
+    return str(id)   
+```
+
+An to get it back you would need:
+
+```	
+def get(self, id):
+    unpickled_obj = self.r.hget(self.hash_name, str(id))
+    obj = pickle.loads(unpickled_obj)
+    return obj
+```
+
+Pickling seemed like a good solution at first, but when we needed to nest remote objects, e.g. put *TangoMachine*s into the remote *TangoQueue*, we ran into the problem of `self.r`(type of *redis.StrictRedis*) not being serializable. To get around this, we defined custom __getstate__ and __setstate__ methods that would not include self.r attribute in the serialized string, but set it back when the object is deserialized.
+
+```
+def __setstate__(self, dict):
+    self.__db= getRedisConnection()
+    self.__dict__.update(dict)
+```
+
+Details of the implementation can be found on [TangoObjects module](https://github.com/autolab/Tango/blob/master/tangoObjects.py).
+
+
+## Producer Consumer Model ##
+
+
+So far we have covered how Tango receives grading jobs and stores them. The second most important part of the process is to read from the queue and distribute the jobs to the available containers(cluster of Virtual Machine or Docker). In the initial architecture, the consumer was part of the HTTP server process. A single Tango process would both produce and consume the jobs.
+
+Having a shared queue gave us the ability to run multiple *producers* as explained in the previous section. However, we only needed one *consumer* to distribute the jobs.
+
+Therefore, we decided to separate the consumer from the HTTP server (producer) as a standalone process.
+
+![Tango with Prod/Com Model]({{site-url}}/assets/redis3.png)
+
+_Diagram 3: Tango with Prod/Com processes and Persistent Memory Model_
+
+
+The advantage of this architecture is that we can launch arbitrary number of HTTP servers as we receive more load. There would be only one consumer process which is responsible for reading from the queue and assigning the jobs. This process gives us a lot of flexibility because it can be run on any node, be stopped and migrated to a different machine at any time.
```

Comment:
- is that we can launch __an__ arbitrary number
- on any node, __as well as__ be stopped and migrated